### PR TITLE
Convert JavaScript To TypeScript

### LINKED ISSUE #18: TS7016: Could not find a declaration file for module '@hotwired/turbo-rails'
Ran into this error while upgrading a Rails 6 app that uses TypeScript and Turbolinks to the new Turbo. Any ideas? [No results on DefinitelyTyped yet.](https://www.typescriptlang.org/dt/search?search=hotwire)

```
ERROR in app/javascript/packs/application.ts:10:30
TS7016: Could not find a declaration file for module '@hotwired/turbo-rails'. '/myapp/node_modules/@hotwired/turbo-rails/app/assets/javascripts/turbo.js' implicitly has an 'any' type.
Try `npm i --save-dev @types/hotwired__turbo-rails` if it exists or add a new declaration (.d.ts) file containing `declare module '@hotwired/turbo-rails';`
  8 | 
> 9 | import { Turbo, cable } from "@hotwired/turbo-rails";
```

### PR DESCRIPTION
**Pull request by kzkn**
For fix #18

### CODE REVIEW

File: app/javascript/turbo/cable.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export async function getConsumer() {
+  return consumer || createConsumer().then(setConsumer)
+}
+
+export function setConsumer(newConsumer: Consumer) {
+  return consumer = newConsumer
+}
+
+export async function createConsumer() {
+  const { createConsumer } = await import(/* webpackChunkName: "actioncable" */ "@rails/actioncable/src")
```
Comment (kzkn) [inline app/javascript/turbo/cable.ts:19]: I could't set `"strict": true` in tsconfig.json because I couldn't resolve the type error on this line.

Comment (kzkn) [reply app/javascript/turbo/cable.ts]: I have fixed the type error.

**Comment by seanpdoyle** [comment]
@kzkn thank you for taking this one on!

As the original author of `app/javascript/turbo/fetch_requests.ts`, I thought it'd be good to pull this down locally and test out the types. How do you feel about making some of the following changes?

```diff
diff --git a/app/javascript/turbo/fetch_requests.ts b/app/javascript/turbo/fetch_requests.ts
index a62e0a4..c78daf3 100644
--- a/app/javascript/turbo/fetch_requests.ts
+++ b/app/javascript/turbo/fetch_requests.ts
@@ -1,36 +1,44 @@
 import type { TurboBeforeFetchRequestEvent, TurboSubmitStartEvent } from '@hotwired/turbo'
 
-export function encodeMethodIntoRequestBody(event: Event) {
-  let ev = event as TurboBeforeFetchRequestEvent
-  if (ev.target instanceof HTMLFormElement) {
-    const { target: form, detail: { fetchOptions } } = ev
-
-    form.addEventListener("turbo:submit-start", ev => {
-      const { detail: { formSubmission: { submitter } } } = ev as TurboSubmitStartEvent
-      const method = detectSubmitMethod(form, submitter, fetchOptions) || ""
+export function encodeMethodIntoRequestBody({ target: form, detail: { fetchOptions } }: TurboBeforeFetchRequestEvent) {
+  if (form instanceof HTMLFormElement) {
+    form.addEventListener("turbo:submit-start", <EventListener>(({ detail: { formSubmission: { submitter } } }: TurboSubmitStartEvent) => {
+      const method = detectSubmitMethod(form, submitter, fetchOptions)
 
       if (!/get/i.test(method)) {
-        const body = fetchOptions.body as FormData | undefined
-        if (/post/i.test(method)) {
-          body?.delete("_method")
-        } else {
-          body?.set("_method", method)
-        }
+        const body = fetchOptions.body
+
+        if (isFormData(body)) {
+          if (/post/i.test(method)) {
+            body.delete("_method")
+          } else {
+            body.set("_method", method)
+          }
 
-        fetchOptions.method = "post"
+          fetchOptions.method = "post"
+        }
       }
-    }, { once: true })
+    }), { once: true })
   }
 }
 
-function detectSubmitMethod(form: HTMLElement, submitter: HTMLElement | undefined, fetchOptions: RequestInit): string | null {
-  const _submitter = submitter as HTMLInputElement | HTMLButtonElement | undefined
-  const body = fetchOptions.body as FormData | undefined
-  if (_submitter && _submitter.formMethod) {
-    return _submitter.formMethod
-  } else if (body?.get("_method")) {
-    return body.get("_method") as string | null
+function detectSubmitMethod(form: HTMLElement, submitter: HTMLElement | undefined, fetchOptions: RequestInit): string {
+  const body = fetchOptions.body
+  const method = isFormData(body) ? body.get("_method") : null
+
+  if (overridesMethod(submitter) && submitter.formMethod) {
+    return submitter.formMethod
+  } else if (typeof method == "string") {
+    return method
   } else {
-    return form.getAttribute("method")
+    return form.getAttribute("method") || "get"
   }
 }
+
+function isFormData(body: BodyInit | null | undefined): body is FormData {
+  return body instanceof FormData
+}
+
+function overridesMethod(submitter: HTMLElement | undefined): submitter is HTMLButtonElement | HTMLInputElement {
+  return submitter instanceof HTMLButtonElement || submitter instanceof HTMLInputElement
+}
diff --git a/app/javascript/turbo/index.ts b/app/javascript/turbo/index.ts
index 7a75be5..3e25eea 100644
--- a/app/javascript/turbo/index.ts
+++ b/app/javascript/turbo/index.ts
@@ -8,4 +8,4 @@ export { cable }
 
 import { encodeMethodIntoRequestBody } from "./fetch_requests"
 
-addEventListener("turbo:before-fetch-request", encodeMethodIntoRequestBody)
+addEventListener("turbo:before-fetch-request", <EventListener>(encodeMethodIntoRequestBody))
```

File: app/javascript/turbo/fetch_requests.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import type { TurboBeforeFetchRequestEvent, TurboSubmitStartEvent } from '@hotwired/turbo'
+
+export function encodeMethodIntoRequestBody(event: Event) {
+  let ev = event as TurboBeforeFetchRequestEvent
+  if (ev.target instanceof HTMLFormElement) {
+    const { target: form, detail: { fetchOptions } } = ev
+
+    form.addEventListener("turbo:submit-start", ev => {
```
Comment (marcoroth) [inline app/javascript/turbo/fetch_requests.ts:8]: To add what @seanpdoyle mentioned in the previous comment: I've opened https://github.com/hotwired/turbo/pull/800 to help with this.

With that you would be able to do something like:
```suggestion
export function encodeMethodIntoRequestBody(event: TurboBeforeFetchRequestEvent) {
  if (event.target instanceof HTMLFormElement) {
    const { target: form, detail: { fetchOptions } } = event

    form.addEventListener("turbo:submit-start", (ev: TurboSubmitStartEvent) => {
```

**Review by marcoroth — APPROVED** [review]
Thanks for starting this one @kzkn!

This looks good for me so far, I've left two small improvements, but they are non-blocking.

File: app/javascript/turbo/fetch_requests.ts (hunk 2 above)
Comment (kzkn) [reply app/javascript/turbo/fetch_requests.ts:8]: It looks good to me. I look forward to seeing the pullreq merged 😀

**Comment by kzkn** [comment]
@seanpdoyle 
Thanks for the comment.
I had a hard time sorting this part out. Your code looks better than mine :smile: 
I rewrote it based on your comments and the code written in #403.

**Comment by kzkn** [comment]
I have rebased to follow #410

**Comment by brunoprietog** [comment]
Since Turbo switched to JavaScript, I think we can close this

**Pull request closed without merging**